### ISSUE TITLE
Error deserializing a serialized contract, modified on-the-fly

### ISSUE THREAD

**Issue by nielstron** [Hyperion-BT/helios#53]
So what I tried to do is deserializeFromBytes a contract, apply three Helios values (Int, ByteString) and then serialized again. This produces invalid cbor (the following):

```cbor
591fc7010000333222223253335734a666666ae900084cdc39aab9d375400490000a5014a029405280999800a4c0060042660020060046664444446493191919999998009bae008375c00e6eb401892600400300253300148901220013263357389201144e616d654572726f723a2076616c696461746f7200498c8c8c8894ccd5cd19b8f002488101220011003133004002001222222232498c8c94ccd5cd1919b8732337046466e04c94cc005221010d0013263357389201154e616d654572726f723a20616c6c5f6c6f636b656400498004c94cc0052201010e0013263357389201174e616d654572726f723a20616c6c5f756e6c6f636b656400498004004c94cc005221011200132633573892011a4e616d654572726f723a207772617070696e675f666163746f7200498004004c94cc005221010f0013263357389201154e616d654572726f723a20616c6c5f6d696e7465640049800400440044c98cd5ce1919b96324912b57726f6e6720616d6f756e74206f6620746f6b656e73206d696e7465642c20646966666572656e63653a2000001323332533001489010500132633573892010e4e616d654572726f723a2073747200498004c8cdc09919b82323370264a660029101010d0013263357389201154e616d654572726f723a20616c6c5f6c6f636b656400498004c94cc0052201010e0013263357389201174e616d654572726f723a20616c6c5f756e6c6f636b656400498004004c94cc005221011200132633573892011a4e616d654572726f723a207772617070696e675f666163746f7200498004004c94cc005221010f0013263357389201154e616d654572726f723a20616c6c5f6d696e7465640049800400400400400526323200232333253300148901060013263357389201104e616d654572726f723a207072696e7400498004c8cdcb192481086d696e7465643a2000001323332533001489010500132633573892010e4e616d654572726f723a2073747200498004c94cc0052201010f0013263357389201154e616d654572726f723a20616c6c5f6d696e74656400498004004004004004004c8c8008c8ccc94cc005220101060013263357389201104e616d654572726f723a207072696e7400498004c8cdcb192481146c6f636b656420617420636f6e74726163743a2000001323332533001489010500132633573892010e4e616d654572726f723a2073747200498004c94cc0052201010d0013263357389201154e616d654572726f723a20616c6c5f6c6f636b656400498004004004004004004c8c8008c8ccc94cc005220101060013263357389201104e616d654572726f723a207072696e7400498004c8cdcb19248118756e6c6f636b65642066726f6d20636f6e74726163743a2000001323332533001489010500132633573892010e4e616d654572726f723a2073747200498004c94cc0052201010e0013263357389201174e616d654572726f723a20616c6c5f756e6c6f636b656400498004004004004004004c8c8894ccd5cd19b8f0024881010f0011003133004002001323333232222375a6aae78cc8c8c004004894ccd5cd1aba30011337606ea4018dd40028a999ab9a30033574200226ae840044cc008008d5d1000919baf37520086aae74004010c8cccc8c8888dd59aab9e33232300100122533357346ae8c0044cdd81ba9006374c00a2a666ae68c00cd5d080089aba100113300200235744002466ebcdd48021aab9d0010043232375666646002002444a666ae68d5d1800899319ab9c4910a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40106aae78dd50009919199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a45010300132633573892010e4e616d654572726f723a2063747800498004004004004c94cc00522101130013263357389201124e616d654572726f723a206f776e5f70696400498004c8cd5d019bb037526491010000137506490000009bb2498004004004004c8cdc5192450177000013253300148901110013263357389201154e616d654572726f723a20746f6b656e5f6e616d6500498004004c920000010010013232225333573466e3c0092201010e001100313300400200132333332533001489011a0013263357389201344e616d654572726f723a20616c6c5f746f6b656e735f756e6c6f636b65645f66726f6d5f636f6e74726163745f6164647265737300498004c8c8cc8c8c004004894ccd5cd1aba30011376293099aba0300335742002660040046ae880048004dd6199918008009112999ab9a35746002264c66ae7124010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd50009919199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a45010300132633573892010e4e616d654572726f723a2063747800498004004004c94cc0052201010c0013263357389201134e616d654572726f723a206f776e5f6164647200498004c94cc005220101000013263357389201104e616d654572726f723a20746f6b656e00498004004004c8c8894ccd5cd19b8f002489010d001100313300400200132333332533001489011f0013263357389201304e616d654572726f723a20616c6c5f746f6b656e735f6c6f636b65645f61745f636f6e74726163745f6164647265737300498004c8c8cc8c8c004004894ccd5cd1aba30011376293099aba0300335742002660040046ae880048004dd6199918008009112999ab9a35746002264c66ae7124010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40086aae78dd50009919199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a45010300132633573892010e4e616d654572726f723a2063747800498004004004c94cc0052201010c0013263357389201134e616d654572726f723a206f776e5f6164647200498004c94cc005220101000013263357389201104e616d654572726f723a20746f6b656e00498004004004c8c8894ccd5cd19b8f002489010000110031330040020013233332533001488101190013263357389201104e616d654572726f723a20546f6b656e00498004c94cc0052201012100132633573892011a4e616d654572726f723a20746f6b656e5f706f6c6963795f696400498004c94cc005220101110013263357389201154e616d654572726f723a20746f6b656e5f6e616d6500498004004004c94ccd5cd1919b87323235573a6ea8004c94cc00522101040013263357389201124e616d654572726f723a20707572706f736500498004004c920000010011323232225333573466e3c0092201011300110031330040020013232375c66646002002444a666ae68d5d1800899319ab9c49010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a4501040013263357389201124e616d654572726f723a20707572706f736500498004004c8c8894ccd5cd19b8f0024881010c0011003133004002001323332533001489011e0013263357389201164e616d654572726f723a206f776e5f6164647265737300498004c8c8dd7199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a4501040013263357389201124e616d654572726f723a20707572706f7365004980040040040040040044c8c94ccd5cd1919b87323235573a6ea8004c94cc00522101040013263357389201124e616d654572726f723a20707572706f736500498004004c920020010011323232225333573466e3c0092201010c0011003133004002001323233323001001222533357346ae8c0044c98cd5ce24810a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a4501140013263357389201134e616d654572726f723a206f776e5f7574786f00498004004c8c8894ccd5cd19b8f002488101130011003133004002001323332533001489010a0013263357389201184e616d654572726f723a206f776e5f706f6c6963795f696400498004c94cc005220101140013263357389201134e616d654572726f723a206f776e5f7574786f00498004004004c8c8894ccd5cd19b8f002489011400110031330040020013233332533001488101090013263357389201194e616d654572726f723a206f776e5f7370656e745f7574786f00498004c8c8cc8c8c004004894ccd5cd1aba30011376293099aba0300335742002660040046ae880048004dd6199918008009112999ab9a35746002264c66ae7124010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd50009919199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a45010300132633573892010e4e616d654572726f723a2063747800498004004004c94cc005220101040013263357389201124e616d654572726f723a20707572706f7365004980040040040040044c8c94ccd5cd192500011001132633573864920117496e636f727265637420707572706f736520676976656e00001498004004004004c8c8894ccd5cd19b8f00248901040011003133004002001323233323001001222533357346ae8c0044c98cd5ce24810a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40046aae78dd500099299800a45010300132633573892010e4e616d654572726f723a2063747800498004004004cccccc888888894ccd5cd19b8f0024890103001100315333573466e3c00922010124001100415333573466e3c00922010123001100515333573466e3c00922010112001100615333573466e3c00922010111001100715333573466e3c0092201012100110081330090020010070060050040030023232225333573466e3c009221011f0011003133004002001222232533001489010100132633573892010e4e616d654572726f723a2072657300498c8c8ccc8c8c0040048894ccd5cd1aba30021001133300300335744004660080026ae8400888c8c94ccd5cd1919baf323233323001001222533357346ae8c0044c98cd5ce2490a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a45010b00132633573892010e4e616d654572726f723a2074786f00498004004c94cc00522101080013263357389201124e616d654572726f723a2061646472657373004980040044c8c94ccd5cd1919baf323233323001001222533357346ae8c0044c98cd5ce2490a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40086aae78dd500099299800a45010b00132633573892010e4e616d654572726f723a2074786f00498004004c8ccc94cc005221011800132633573892011a4e616d654572726f723a20536f6d654f7574707574446174756d00498004dd49924500001001001001100113263357386492012e446f6573206e6f742061747461636820636f727265637420646174756d20746f20736372697074206f757470757400001498c8c8894ccd5cd19b8f00248901010011003133004002001323370064a660029101010100132633573892010e4e616d654572726f723a2072657300498004c8cccc8c8888dd69aab9e33232300100122533357346ae8c0044cdd81ba9006375000a2a666ae68c00cd5d080089aba100113300200235744002466ebcdd48021aab9d00100432333323222237566aae78cc8c8c004004894ccd5cd1aba30011337606ea4018dd30028a999ab9a30033574200226ae840044cc008008d5d1000919baf37520086aae74004010c8c8dd5999918008009112999ab9a35746002264c66ae7124010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40046aae78dd500099299800a45010b00132633573892010e4e616d654572726f723a2074786f00498004004004c8c8dd7199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a4501000013263357389201104e616d654572726f723a20746f6b656e00498004004c8cd5d019bb03752649110000137506490000009bb2498004004004004c8c8dd7199918008009112999ab9a35746002264c66ae7124010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40046aae78dd500099299800a4501000013263357389201104e616d654572726f723a20746f6b656e00498004004c9200000100100100100100113200100100132225333573466e3c009221010b001100313300500200100132533001488101200013263357389201114e616d654572726f723a2074786f75747300498004004c8c8894ccd5cd19b8f0024881010100110031330040020013248000004004ccc888894ccd5cd19b8f0024890100001100315333573466e3c00922010108001100415333573466e3c0092201012000110051330060020010040030023232225333573466e3c0092201011e0011003133004002001223233332533001488101170013263357389201124e616d654572726f723a204164647265737300498004c8ccc94cc0052201011500132633573892011b4e616d654572726f723a2053637269707443726564656e7469616c00498004c94cc0052201010a0013263357389201184e616d654572726f723a206f776e5f706f6c6963795f696400498004004004c8cc94cc005221011600132633573892011e4e616d654572726f723a204e6f5374616b696e6743726564656e7469616c00498004004004004c8004c8894ccd5cd19b8f002489010a00110031330040020010023232225333573466e3c0092201010a00110031330040020012232533001489011d0013263357389201144e616d654572726f723a20706f6c6963795f696400498c8c94ccd5cd1919b87323235573a6ea8004c94cc0052201011000132633573892010f4e616d654572726f723a206372656400498004004c920020010011323232225333573466e3c0092201011d00110031330040020013232375c66646002002444a666ae68d5d1800899319ab9c49010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a45011000132633573892010f4e616d654572726f723a2063726564004980040040040044c8004004c8c8894ccd5cd19b8f00248901100011003133004002001323233323001001222533357346ae8c0044c98cd5ce24810a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd50009919199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a4501090013263357389201194e616d654572726f723a206f776e5f7370656e745f7574786f00498004004004004c8894ccd5cd19b8f0024881010900110031330040020010023232225333573466e3c009220101090011003133004002001222325330014881011c0013263357389201144e616d654572726f723a206f776e5f74786f757400498c8c8ccc8c8c0040048894ccd5cd1aba30021001133300300335744004660080026ae8400888c8c94ccd5cd1919baf323233323001001222533357346ae8c0044c98cd5ce2490a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a45010200132633573892010e4e616d654572726f723a2074786900498004004c8c8ccc8c0040048894ccd5cd1aba3001132633573892010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a45011b00132633573892010c4e616d654572726f723a2070004980040040044c8c8c8894ccd5cd19b8f0024881011c0011003133004002001323233323001001222533357346ae8c0044c98cd5ce24810a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40046aae78dd500099299800a45010200132633573892010e4e616d654572726f723a20747869004980040040040044c8004004004c8894ccd5cd19b8f00248810102001100313300500200100132533001488101070013263357389201104e616d654572726f723a207478696e7300498004004004cc88894ccd5cd19b8f002489011b001100315333573466e3c0092201010700110041330050020010030023232225333573466e3c009221011a0011003133004002001222232533001489010100132633573892010e4e616d654572726f723a2072657300498c8c8ccc8c8c0040048894ccd5cd1aba30021001133300300335744004660080026ae8400888c8c94ccd5cd1919baf323233323001001222533357346ae8c0044c98cd5ce2490a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd50009919199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40046aae78dd500099299800a45010200132633573892010e4e616d654572726f723a2074786900498004004004c94cc005220101080013263357389201124e616d654572726f723a2061646472657373004980040044c8c8c8894ccd5cd19b8f00248901010011003133004002001323370064a660029101010100132633573892010e4e616d654572726f723a2072657300498004c8cccc8c8888dd69aab9e33232300100122533357346ae8c0044cdd81ba9006375000a2a666ae68c00cd5d080089aba100113300200235744002466ebcdd48021aab9d00100432333323222237566aae78cc8c8c004004894ccd5cd1aba30011337606ea4018dd30028a999ab9a30033574200226ae840044cc008008d5d1000919baf37520086aae74004010c8c8dd5999918008009112999ab9a35746002264c66ae7124010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40046aae78dd50009919199918008009112999ab9a35746002264c66ae712410a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40046aae78dd500099299800a45010200132633573892010e4e616d654572726f723a2074786900498004004004004c8c8dd7199918008009112999ab9a35746002264c66ae7124010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40006aae78dd500099299800a4501000013263357389201104e616d654572726f723a20746f6b656e00498004004c8cd5d019bb03752649110000137506490000009bb2498004004004004c8c8dd7199918008009112999ab9a35746002264c66ae7124010a496e6465784572726f720049854ccd5cd19b87002480004d5d0800899980180199b8100248008d5d1000a40046aae78dd500099299800a4501000013263357389201104e616d654572726f723a20746f6b656e00498004004c9200000100100100100100113200100100132225333573466e3c0092210102001100313300500200100132533001488101070013263357389201104e616d654572726f723a207478696e7300498004004c8c8894ccd5cd19b8f0024881010100110031330040020013248000004004ccc888894ccd5cd19b8f0024890100001100315333573466e3c00922010108001100415333573466e3c0092201010700110051330060020010040030023232225333573466e3c0092201011900110031330040020012223374a900019aba0375200666ae80dd48011bb1498c8c8894ccd5cd19b8f002488101180011003133004002001223374a900219aba00023762931919112999ab9a3371e004911011700110031330040020012223374a900019aba00033357400046ec5263232225333573466e3c0092210116001100313300400200123374a90011bb1498c8c8894ccd5cd19b8f00248901150011003133004002001223374a900119aba037520046ec5263232225333573466e3c00922101060011003133004002001322233573800466e9520003762930009919112999ab9a3371e004910101050011003133004002001322237326464a666ae68cdc3802240002910101300015333573466e200112000133716902d180099b81480000104c0040108ccc8c8c0040048894ccd5cd1aba30021001133300300335744004660080026ae8400888cdc58008011801000a4500323001001225333573466e24005200014bd6099aba03370066e18005201448180cc008008cdc1800a40280020024400200a0080069810105004c01074647726f777468004c011e581c4fd0d998dc0700ca6ef89fafff05fbf523a3a25c1fc8314bf7e4d1c20001
```

I tried deserializing in aiken, which appearently works to parse the original contract, then fails reading the applications

```
$aiken uplc unflat script.cbor --cbor-hex
Error: 
  × Parse error: So far we parsed
  │ 
  │ [ [ [ (lam i(lam i(lam i(lam i(lam i[ (lam i(force [ [ [ (force (builtin
  │ ifThenElse)) (force [ [ [ [ [ [ (force (builtin chooseData)) (var i)]
  │ (delay [ [ (builtin equalsInteger) [ (force (force (builtin fstPair)))
  │ [ (builtin unConstrData) (var i)]]] (con integer  .... the contract ...
  │ #30))] (delay (force [ [ [ (force (builtin ifThenElse)) [ [ (builtin
  │ lessThanInteger) (var i)] (con integer
  │ 0)]] (delay [ [ (builtin consByteString) (con integer
  │ 45)] [ (var i) [ [ (builtin subtractInteger) (con integer
  │ 0)] (var i)]]])] (delay [ (var i) (var i)])]))])) (lam i[ [ [ (lam i[ (lam
  │ i[ (var i) (var i)]) (lam i(lam i(lam i(force [ [ [ (force (builtin
  │ ifThenElse)) [ (force (builtin nullList)) (var i)]] (delay (var i))]
  │ (delay [ [ [ (var i) (var i)] [ (force (builtin tailList)) (var i)]]
  │ [ [ (var i) (var i)] [ (force (builtin headList)) (var i)]]])]))))]) (lam
  │ i(lam i[ [ (builtin consByteString) (var i)] (var i)]))] [ (var i) (var
  │ i)]] (con bytestring
  │ #)])]) [ (lam i[ (var i) (var i)]) (lam i(lam i(force [ [ [ (force
  │ (builtin ifThenElse)) [ [ (builtin lessThanEqualsInteger) (var i)] (con
  │ integer
  │ 0)]] (delay (con list
  │ <integer>
  │ [-26]))] parse error])))]]]))) not parsed]]) not parsed]]]]]]]]]]]]]) not
  │ parsed]]])))))) not parsed] not parsed] not parsed]]))))) not parsed] not
  │ parsed] not parsed]
  │ 
  │ and we ran into error: Unknown term constructor tag: 10.
  │ 
  │ Here are the buffer bytes (5 preceding) [01, 4B, D6, 09, 9A, BA, 03, 37,
  │ 00, 66]
  │ 
  │ Buffer position is 8052 and buffer length is 8135
```

I used these command to apply to the script

```js
    const a = new HeliosByteArray(a_raw);
    const b = new HeliosByteArray(b_raw);
    const c = new HeliosInt(c_raw);
    const parameterizedScript = rawScript.apply([c, b, a]);
    const cborScript = JSON.parse(parameterizedScript.serialize()).cborHex;
```

**Comment by nielstron** [comment]
I found a minimal example here
```js
import { ByteArray as HeliosByteArray, Int as HeliosInt, deserializeUplcBytes, hexToBytes, bytesToHex, CborData, textToBytes, ByteArray } from "./helios.js"

let script = deserializeUplcBytes(
    CborData.decodeBytes(
        hexToBytes(
            "4d0100004bd6f7b42f5a205c8261"
        )
    )
)
script = script.apply([new HeliosByteArray("")])
console.log(script)
console.log(script.serialize())
```

this is the same example as contract I used before (the issue does not occur with the contract (program 1.0.0 (lam x x)), so I assume its due to the containers)

**Comment by nielstron** [comment]
I just noticed this already breaks earlier, even without the application

```Js
import { ByteArray as HeliosByteArray, Int as HeliosInt, deserializeUplcBytes, hexToBytes, bytesToHex, CborData, textToBytes, ByteArray } from "./helios.js"

let script = deserializeUplcBytes(
    CborData.decodeBytes(
        hexToBytes(
            "4d0100004bd6f7b42f5a205c8261"
        )
    )
)
// script = script.apply([new HeliosByteArray("")])
console.log(script)
console.log(script.serialize())
```

is parsed by aiken into 

```
(program
  1.0.0
  (con list<bool> [])
)
```

but should be
```
(program
  1.0.0
  (con list<pair<integer, list<bool>>> [(1, [True,False]),(2, [True])])
)
```

**Comment by nielstron** [comment]
fixed by #54
